framework/src/main/java/org/checkerframework/framework/util/JavaExpressionParseUtil.java

```diff
@@ -569,30 +646,70 @@ public JavaExpression visit(NameExpr expr, JavaExpressionContext context) {
             return null;
         }
 
+        /**
+         * Returns a JavaExpression for the given field.
+         *
+         * <p>{@code isOriginalReceiver} might be false when the user wrote "a.f" but {@code f} is
+         * not declared in the declared type of {@code a}, in which case the this method is called
+         * with {@code receiverExpr}'s type being a supertype of the declared type of {@code a}. In
+         * that case, the resulting FieldAccess's {@code receiver} field is a ThisReference of the
+         * type of {@code receiverExpr}.
+         *
+         * @param fieldElem the field
+         * @param receiverExpr the receiver of the field access; the expression used to access the
+         *     field
+         * @param isOriginalReceiver whether the receiver is the original one written in source code
+         * @return a JavaExpression for the given name
+         */
+        private static FieldAccess getFieldJavaExpression(
+                VariableElement fieldElem,
+                JavaExpression receiverExpr,
+                boolean isOriginalReceiver) {
+            TypeMirror receiverType = receiverExpr.getType();
+
+            TypeMirror fieldType = ElementUtils.getType(fieldElem);
+            if (ElementUtils.isStatic(fieldElem)) {
+                Element classElem = fieldElem.getEnclosingElement();
+                JavaExpression staticClassReceiver = new ClassName(ElementUtils.getType(classElem));
+                return new FieldAccess(staticClassReceiver, fieldType, fieldElem);
+            }
+            JavaExpression locationOfField;
+            if (isOriginalReceiver) {
+                locationOfField = receiverExpr;
+            } else {
+                locationOfField = new ThisReference(receiverType);
```

Comment:
Is `ThisReference` always the right thing?  If the original expression was "a.f" but `f` is declared in a supertype, then might this produce "superthis.f" which could mean something different than "a.f" (because "this" could have its own f field, for example if "this" and "a" have the same type)?

Reply:
As you point out, this code is buggy.  I rewrote it and fix the bugs in a separate pull request: https://github.com/typetools/checker-framework/pull/4388.  (That pull request is also now merged into this.)